File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/JoinGame_Notification.java
package org.emulinker.kaillera.controller.v086.protocol;

import com.google.auto.value.AutoValue;
import com.google.common.base.Strings;
import org.emulinker.kaillera.controller.messaging.MessageFormatException;

@AutoValue
public abstract class JoinGame_Notification extends JoinGame {
  private static final String DESC = "Join Game Notification";

  public static AutoValue_JoinGame_Notification create(
      int messageNumber,
      int gameId,
      int val1,
      String username,
      long ping,
      int userId,
      byte connectionType)
      throws MessageFormatException {
    V086Message.validateMessageNumber(messageNumber, DESC);

    if (gameId < 0 || gameId > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: gameID out of acceptable range: " + gameId);
    }

    if (ping < 0 || ping > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: ping out of acceptable range: " + ping);
    }

    if (userId < 0 || userId > 0xFFFF) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: userID out of acceptable range: " + userId);
    }

    if (connectionType < 1 || connectionType > 6) {
      throw new MessageFormatException(
          "Invalid " + DESC + " format: connectionType out of acceptable range: " + connectionType);
    }

    if (Strings.isNullOrEmpty(username)) {
      throw new MessageFormatException("Invalid " + DESC + " format: userName.length() == 0");
    }
    return new AutoValue_JoinGame_Notification(
        messageNumber, JoinGame.ID, gameId, val1, username, ping, userId, connectionType);
  }

  @Override
  public String description() {
    return DESC;
  }
}

File: emulinker/src/main/java/org/emulinker/kaillera/controller/v086/protocol/JoinGame_Notification.kt
package org.emulinker.kaillera.controller.v086.protocol

import org.emulinker.kaillera.controller.messaging.MessageFormatException
import org.emulinker.kaillera.controller.v086.protocol.V086Message.Companion.validateMessageNumber
import org.emulinker.kaillera.model.ConnectionType

data class JoinGame_Notification
    @Throws(MessageFormatException::class)
    constructor(
        override val messageNumber: Int,
        override val gameId: Int,
        override val val1: Int,
        override val username: String,
        override val ping: Long,
        override val userId: Int,
        override val connectionType: ConnectionType
    ) : JoinGame() {

  override val messageId = ID

  init {
    validateMessageNumber(messageNumber)
    require(gameId in 0..0xFFFF) { "gameID out of acceptable range: $gameId" }
    require(ping in 0..0xFFFF) { "ping out of acceptable range: $ping" }
    require(userId in 0..0xFFFF) { "UserID out of acceptable range: $userId" }
    require(username.isNotBlank()) { "Username cannot be empty" }
  }
}
